packages/core/src/electron-node/electron-node-backend-module.ts

```diff
@@ -0,0 +1,114 @@
+        // We'll multiplex messages over a Node IPC connection and talk JSON-RPC over the channels.
+        bind(ConnectionMultiplexer)
+            .toDynamicValue(ctx => {
+                const transformer = ctx.container.get(ConnectionTransformer);
+                const nodeIpcConnectionFactory = ctx.container.get(NodeIpcConnectionFactory);
+                const deferredConnectionFactory = ctx.container.get(DeferredConnectionFactory);
+                const parentIpc = nodeIpcConnectionFactory(process);
+                const sharedIpc: AnyConnection = transformer(parentIpc, {
+                    decode: (message, emit) => {
+                        if (typeof message === 'object' && THEIA_ELECTRON_IPC_CHANNEL_NAME in message) {
+                            emit(message[THEIA_ELECTRON_IPC_CHANNEL_NAME]);
+                        }
+                    },
+                    encode: (message, write) => {
+                        write({ [THEIA_ELECTRON_IPC_CHANNEL_NAME]: message });
+                    }
+                });
+                const deferredConnection = deferredConnectionFactory(waitForRemote(sharedIpc));
+                return ctx.container.get(DefaultConnectionMultiplexer).initialize(deferredConnection);
+            })
+            .inSingletonScope()
+            .whenTargetNamed(ElectronMainAndBackend);
+        bind(BackendApplicationContribution)
+            .toDynamicValue(ctx => ({
+                initialize(): void {
+                    const multiplexer = ctx.container.getNamed(ConnectionMultiplexer, ElectronMainAndBackend);
+                    const serviceProvider = ctx.container.getNamed(ServiceProvider, ElectronMainAndBackend);
+                    const jsonRpcConnectionFactory = ctx.container.get(JsonRpcConnectionFactory);
+                    const rpcProxying = ctx.container.get(RpcProxying);
+                    multiplexer.listen(({ serviceId, serviceParams }, accept, next) => {
+                        const service = serviceProvider.getService(serviceId, serviceParams);
+                        if (service) {
+                            rpcProxying.serve(service, jsonRpcConnectionFactory(accept()));
+                        } else {
+                            next();
+                        }
+                    });
+                }
+            }))
+            .inSingletonScope();
+        bind(ProxyProvider)
+            .toDynamicValue(ctx => {
+                const multiplexer = ctx.container.getNamed(ConnectionMultiplexer, ElectronMainAndBackend);
+                const jsonRpcConnectionFactory = ctx.container.get(JsonRpcConnectionFactory);
+                return ctx.container.get(DefaultRpcProxyProvider).initialize(
+                    (serviceId, serviceParams) => jsonRpcConnectionFactory(multiplexer.open({ serviceId, serviceParams }))
+                );
+            })
+            .inSingletonScope()
+            .whenTargetNamed(ElectronMainAndBackend);
```

Comment:
Maybe I'm not understanding the concepts here yet, but it just seem I need to understand too many interlocking concepts at once here: If I think about what we're trying to achieve, it's not that complicated:

1. The system establishes a number of well know point-to-point connections at startup, for example front end <=> back end or front end <=> plugin host or back end <=> file system change watcher
2. Some clients want to bind service objects whose lifecycle it bound to those connections when the connections are established. 
3. Some clients bind service proxies to those connections. 

When we establish a connection, we establish a service scope for those services. The connection being established is put inside this scope and allow any listeners to react to the connection.
We can bind a rpc protocol factory that will create an rpc protocol instance in this new scope. it will be bound to the connection. 
We need a contribution point that allows clients to bind service object factories and client proxy factories inside this scope. There needs to be an initialization protocol between clients and services that is two-phase: only after full initialization are remote calls allowed.
The services are bound to a named endpoints, for example: "BackEndFrontEnd" or "PluginHost". Note that multiple instances of the same endpoint may be established (such as with plugin hosts).